### PR TITLE
add integration tests for basic, sso, and openid auth

### LINKED ISSUE #163: Integration Testing - Jetty
**Integration Testing - Jetty**

Integration testing CWMS AAA login and OpenDCS API Authorization on Jetty.

### LINKED ISSUE #164: Integration Testing - Tomcat
**Integration Testing - Tomcat**

Integration testing CWMS AAA login and OpenDCS API authorization on Tomcat.

### LINKED ISSUE #165: Integration Testing - Failure Cases
**Integration Testing - Failure Cases**

Integration testing OpenDCS API authorization failures when unauthenticated.

### PR DESCRIPTION
**Pull request by adamkorynta**
## Problem Description

Fixes #163
Fixes #164 
Fixes #165

Add integration tests for running auth against basic, sso, and openid authorization.

## Solution

Used RestAssured to run requests against the API.

## how you tested the change

Change was adding tests

## Where the following done:

- [X] Tests. Check all that apply:
   - [X] Unit tests created or modified that run during ant test.
   - [X] Integration tests created or modified that run during integration testing
         (Formerly called regression tests.)
   - [ ] Test procedure descriptions for manual testing
- [ ] Was relevant documentation updated?
- [X] Were relevant config element (e.g. XML data) updated as appropriate

### CODE REVIEW

File: opendcs-rest-api/src/test/java/org/opendcs/odcsapi/fixtures/TomcatServer.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	 * arg[0] - the CATALINA_BASE directory you've setup
+	 * arg[1] - full path to the war file generated by this build script
+	 * arg[2] - name to use for this instance. See constructor for guidance
+	 *
+	 * @param args standard argument list
+	 */
+	public static void main(String[] args)
+	{
+		String baseDir = args[0];
+		String cdaWar = args[1];
```
Comment (MikeNeilson) [inline opendcs-rest-api/src/test/java/org/opendcs/odcsapi/fixtures/TomcatServer.java]: considering this was clearly taken directly from CDA, I wonder if we should just put it in its own project for convience?

File: opendcs-rest-api/src/test/java/org/opendcs/odcsapi/fixtures/TomcatServer.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+	 * arg[0] - the CATALINA_BASE directory you've setup
+	 * arg[1] - full path to the war file generated by this build script
+	 * arg[2] - name to use for this instance. See constructor for guidance
+	 *
+	 * @param args standard argument list
+	 */
+	public static void main(String[] args)
+	{
+		String baseDir = args[0];
+		String contextName = args[1];
```
Comment (MikeNeilson) [inline opendcs-rest-api/src/test/java/org/opendcs/odcsapi/fixtures/TomcatServer.java:137]: I saw that. :smile:. But considering this was lifted near verbatim from CDA I wonder if we should just make our little "embedded tomcat for test" it's own project.

File: opendcs-rest-api/src/test/java/org/opendcs/odcsapi/sec/BaseIT.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.opendcs.odcsapi.fixtures.TomcatServer;
+import org.slf4j.Logger;
+
+import static io.restassured.RestAssured.given;
+import static org.hamcrest.Matchers.is;
+
+public abstract class BaseIT
```
Comment (MikeNeilson) [inline opendcs-rest-api/src/test/java/org/opendcs/odcsapi/sec/BaseIT.java:31]: The tomcat and database operations would probably work better as an extension than just a base class.

Also possible I've just written way too many extensions and gotten used to the idea.

File: opendcs-rest-api/src/test/resources/tomcat/conf/tomcat-server.xml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        <Engine name="Catalina" defaultHost="localhost">
+            <Host name="localhost" appBase="webapps" unpackWARs="true" autoDeploy="true">
+                <!--
+                SingleSignOn valve, share authentication between web applications
+                        Documentation at: /docs/config/valve.html
+                -->
+                <!--<Valve className="fixtures.TestAuthValve"/>-->
+                <Valve className="org.apache.catalina.authenticator.SingleSignOn"/>
+                <Valve className="org.apache.catalina.valves.ErrorReportValve"
+                    showReport="false"
```
Comment (MikeNeilson) [inline opendcs-rest-api/src/test/resources/tomcat/conf/tomcat-server.xml:47]: These can, and perhaps should, be true for test runs.

**Review by MikeNeilson — COMMENTED** [review]
Current notes.

File: opendcs-rest-api/src/test/java/org/opendcs/odcsapi/fixtures/TomcatServer.java (hunk 1 above)
Comment (adamkorynta) [reply opendcs-rest-api/src/test/java/org/opendcs/odcsapi/fixtures/TomcatServer.java]: I couldn't get it working with the deployed war file like in CDA, having jaxrs jars on the tomcat class path as well as in the war libs sure makes it mad. After several failed attempts at cleaning up the path, I gave up and just left it out. Dunno if there's really enough common code for a shared project, but maybe?

File: opendcs-rest-api/src/test/java/org/opendcs/odcsapi/sec/BaseIT.java (hunk 3 above)
Comment (adamkorynta) [reply opendcs-rest-api/src/test/java/org/opendcs/odcsapi/sec/BaseIT.java]: good idea

File: opendcs-rest-api/src/test/resources/tomcat/conf/tomcat-server.xml (hunk 4 above)
Comment (adamkorynta) [reply opendcs-rest-api/src/test/resources/tomcat/conf/tomcat-server.xml]: updated

File: opendcs-rest-api/src/test/java/org/opendcs/odcsapi/fixtures/TomcatServer.java (hunk 1 above)
Comment (MikeNeilson) [reply opendcs-rest-api/src/test/java/org/opendcs/odcsapi/fixtures/TomcatServer.java]: Fair, I can see there being some contention with jaxrs and tomcat. Not that perhaps there should be but why fight it unless we have to.

File: opendcs-rest-api-jetty/Dockerfile-install-postgres-schema [hunk 5]
```diff
@@ -6,6 +6,7 @@ RUN apt-get install openjdk-17-jre -y
 COPY ./docker-config/ /opendcs/
 COPY ./build/tmp/opendcs-installer.jar /opendcs/opendcs-installer.jar
 COPY ./docker-config/InstallTsdbSchemaDocker.sh /docker-entrypoint-initdb.d/InstallTsdbSchemaDocker.sh
+RUN chmod 777 /docker-entrypoint-initdb.d/InstallTsdbSchemaDocker.sh
```
Comment (wjonassen) [inline opendcs-rest-api-jetty/Dockerfile-install-postgres-schema:9]: Do permissions need to be set here?

Comment (adamkorynta) [reply opendcs-rest-api-jetty/Dockerfile-install-postgres-schema:9]: yeah that was part of the issue running locally. Didn't help when running through GitHub Actions though

Comment (wjonassen) [reply opendcs-rest-api-jetty/Dockerfile-install-postgres-schema:9]: Got it....We can leave it.

File: opendcs-rest-api-jetty/Dockerfile-install-postgres-schema [hunk 6]
```diff
@@ -6,6 +6,7 @@ RUN apt-get install openjdk-17-jre -y
 COPY ./docker-config/ /opendcs/
 COPY ./build/tmp/opendcs-installer.jar /opendcs/opendcs-installer.jar
```
Comment (MikeNeilson) [inline opendcs-rest-api-jetty/Dockerfile-install-postgres-schema:7]: fyi, 7.0.12 has docker images, you could just derive from it and avoid the install.

Comment (MikeNeilson) [reply opendcs-rest-api-jetty/Dockerfile-install-postgres-schema:7]: if you are willing to risk 7.0.13 there is a "migateDatabase" app that will do the schema install, including baseline data.

Comment (MikeNeilson) [reply opendcs-rest-api-jetty/Dockerfile-install-postgres-schema:7]: i just looked at the first line. don't do this. postgres should be it's own service in the compose file. 

mixing things in this way is going to be super unstable.

Comment (adamkorynta) [reply opendcs-rest-api-jetty/Dockerfile-install-postgres-schema:7]: Yeah the plan is to throw this all away and update to the migrateDatabase app

**Review by wjonassen — APPROVED** [review]

**Pull request merged**